libcaf_net/caf/net/actor_shell.hpp

```diff
@@ -0,0 +1,209 @@
+#include "caf/detail/net_export.hpp"
+#include "caf/detail/unordered_flat_map.hpp"
+#include "caf/extend.hpp"
+#include "caf/fwd.hpp"
+#include "caf/intrusive/drr_queue.hpp"
+#include "caf/intrusive/fifo_inbox.hpp"
+#include "caf/local_actor.hpp"
+#include "caf/mixin/requester.hpp"
+#include "caf/mixin/sender.hpp"
+#include "caf/net/fwd.hpp"
+#include "caf/none.hpp"
+#include "caf/policy/normal_messages.hpp"
+
+namespace caf::net {
+
+///
+class CAF_NET_EXPORT actor_shell
+  : public extend<local_actor, actor_shell>::with<mixin::sender,
+                                                  mixin::requester>,
+    public dynamically_typed_actor_base,
+    public non_blocking_actor_base {
+public:
+  // -- friends ----------------------------------------------------------------
+
+  friend class actor_shell_ptr;
+
+  // -- member types -----------------------------------------------------------
+
+  using super
+    = extend<local_actor, actor_shell>::with<mixin::sender, mixin::requester>;
+
+  using signatures = none_t;
+
+  using behavior_type = behavior;
+
+  struct mailbox_policy {
+    using queue_type = intrusive::drr_queue<policy::normal_messages>;
+
+    using deficit_type = policy::normal_messages::deficit_type;
+
+    using mapped_type = policy::normal_messages::mapped_type;
+
+    using unique_pointer = policy::normal_messages::unique_pointer;
+  };
+
+  using mailbox_type = intrusive::fifo_inbox<mailbox_policy>;
+
+  using fallback_handler = unique_callback_ptr<result<message>(message&)>;
+
+  // -- constructors, destructors, and assignment operators --------------------
+
+  actor_shell(actor_config& cfg, socket_manager* owner);
+
+  ~actor_shell() override;
+
+  // -- state modifiers --------------------------------------------------------
+
+  /// Detaches the shell from its owner and closes the mailbox.
+  void quit(error reason);
+
+  /// Overrides the callbacks for incoming messages.
+  template <class... Fs>
+  void set_behavior(Fs... fs) {
+    bhvr_ = behavior{std::move(fs)...};
+  }
+
+  /// Overrides the default handler for unexpected messages.
+  template <class F>
+  void set_fallback(F f) {
+    fallback_ = make_type_erased_callback(std::move(f));
+  }
+
+  // -- mailbox access ---------------------------------------------------------
+
+  auto& mailbox() noexcept {
+    return mailbox_;
+  }
+
+  /// Dequeues and returns the next message from the mailbox or returns
+  /// `nullptr` if the mailbox is empty.
+  mailbox_element_ptr next_message();
+
+  /// Tries to put the mailbox into the `blocked` state, causing the next
+  /// enqueue to register the owning socket manager for write events.
+  bool try_block_mailbox();
+
+  // -- message processing -----------------------------------------------------
+
+  /// Dequeues and processes the next message from the mailbox.
+  /// @returns `true` if a message was dequeued and process, `false` if the
+  ///          mailbox was empty.
+  bool consume_message();
```

Comment:
This is not just dequeueing. The shell has a behavior plus default handler that process the message. We're always calling this *consume*, e.g., https://github.com/actor-framework/actor-framework/blob/master/libcaf_core/caf/scheduled_actor.hpp#L428. Consume has different meaning in different contexts, so I don't see a problem TBH and would rather stay consistent.

Reply:
I can understand that argumentation, however I think the semantics of the pointed out function in `scheduled_actor` is different from this case. `consume` in the `scheduled_actor` actually *only* consumes a `mailbox_element` which is passed as argument, while in this case the function dequeues the message from its own queue.

Reply:
> Just to be clear: The function is supposed to handle write events right?

No, I believe you are mixing some contexts. This function takes the next message from the queue and consumes it. That's it. How and when a socket manager would call this function is a separate talking point.

The function in `scheduled_actor` can get away with using an unqualified `consume` as function name, exactly because it takes an argument. That's why the function is called `consume_message` here. The message has to come from somewhere, and for actors the natural assumption is: from the mailbox. I think more descriptive names like `dequeue_and_consume_message` only make the API unwieldy and don't add value.

Reply:
The problem is that I was viewing this component as another layer within the stack. Comparable to a consumer like BASP. I think with this it is safe to conclude this discussion.
I'm fine with the naming of this function 👍